crates/observe/src/future.rs

```diff
@@ -26,24 +26,36 @@ pin_project! {
         #[pin]
         inner: T,
         label: String,
-        timer: Option<prometheus::HistogramTimer>,
+        state: State,
     }
 }
 
+#[derive(Debug)]
+enum State {
+    NeverPolled,
+    Running(prometheus::HistogramTimer),
+    Done,
+}
+
 impl<T: Future> Future for Measurable<T> {
     type Output = T::Output;
 
     fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
         let this = self.project();
-        if this.timer.is_none() {
-            *this.timer = Some(
+        if matches!(this.state, State::NeverPolled) {
+            *this.state = State::Running(
                 Metrics::get()
                     .future_execution_times
                     .with_label_values(&[this.label])
                     .start_timer(),
             );
         }
-        this.inner.poll(cx)
+        let result = this.inner.poll(cx);
+        if let Poll::Ready(_) = result {
+            // Dropping the timer will record the execution time.
+            *this.state = State::Done;
+        }
+        result
     }
 }
```

Comment:
Tiny uber-nit I just thought about, consider adding a:

```rust
impl<T> FusedFuture for Measurable<T>
where
    T: FusedFuture
{}
```

Fused futures have nice properties that they can be `poll`-ed after they return `Poll::Ready`, and some future combinators require this. Without adding an implementation of this marker trait, you may need to `.measure("").fused()` to use it in those cases even in situations where it wouldn't be needed.

Very tiny comment, and fine to ignore, more about "being a good futures citizen".